src/common/libterminus/pty.c

```diff
@@ -646,24 +646,34 @@ int pty_data_unpack (const flux_msg_t *msg,
 {
     const char *encoding = NULL;
     void *data;
-    int len;
+    size_t len;
+    json_t *o;
+    json_error_t error;
 
-    if (flux_msg_unpack (msg,
-                         "{s:s s?s}",
-                         "data", &data,
-                         "encoding", &encoding) < 0)
+    /* Note: Use json_unpack_ex(3) in order to pass the JSON_ALLOW_NUL flag.
+     *  This is necessary since pty data packed with 's#' may contain NUL
+     *  NUL characters, especially in the case of ^@/Ctrl-Space/etc, which
+     *  is encoded as a NUL character (empty string with len=1).
+     */
+    if (flux_msg_unpack (msg, "o", &o) < 0
+        || json_unpack_ex (o,
+                           NULL,
+                           JSON_ALLOW_NUL,
+                           "{s:s% s?s}",
+                           "data", &data, &len,
+                           "encoding", &encoding) < 0) {
+        errno = EPROTO;
```

Comment:
is it possible `flux_msg_unpack()` could return errno != EPROTO, like ENOSYS or something?  Dunno if want to split the if checks up.

Reply:
This is the server side, so ENOSYS not possible.

Reply:
Ok, I split those up. Good suggestion.